### PR TITLE
Add babel-plugin-extract-react-types

### PR DESCRIPTION
**Pull request by mitchellhamilton**
This adds a babel plugin which runs extract-react-types.

I wanted an API to get prop types without having to use the webpack loader syntax so this plugin gets the types for all the exports of a module which are React components(of course it might add types to things which aren't actually React components since you can't know whether a function is a component or not but I don't think it'll really be a problem) and assigns a property `___types` with the types.

I've added another component to pretty-proptypes called ComponentProps which accepts a component and uses the types from the component rather than directly accepting the props. I'm not totally sure about this, maybe the Props component should accept a component prop itself(that raises the question of what happens when you pass both props and a component), maybe there's a better name for ComponentProps?

@Noviny I've added you to the package on npm.

### CODE REVIEW

**Comment by Noviny** [comment]
Hey, sorry this has taken me a while to get back to. Got sick 🤒 and wasn't up to reviewing it. Still looking at code, but wanted to work on the API design bit.

Typing things that aren't a component is probably a feature not a bug (we likely want to move to be able to just show signatures for functions anyway)

I'd rather have one component over two from pretty propTypes.

I think I'd be fine just having:

```
import Props from 'pretty-proptypes';
import someComponent from './somewhere';

<Props
    // This accepts pre-generated data from extract-react-types. You can use the `extract-react-types-loader` inline if you want to keep this simple.
    props={PropInfo}
    // This accepts a react component, and will display its props. This will only work if you have configured `babel-plugin-extract-react-types`. If both are provided, the component will be displayed.
    component={SomeComponent}
/>
```

The `props` prop is a useful fallback if you say wanted to legit write the data to json and read it from somewhere else, instead of making it part of the build.

With regards to overriding I think I would have:

Check if `component` data is sensible (object with kinds), if not, fall back to props but print an error. This lets people provide a fallback if they feel they need to.

File: packages/babel-plugin-extract-react-types/index.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+const { findExportedComponents } = require('extract-react-types');
+
+module.exports = babel => {
+  let t = babel.types;
+  return {
+    visitor: {
+      Program(programPath, state) {
+        let typeSystem = state.file.opts.parserOpts.plugins
+          .map(plugin => (Array.isArray(plugin) ? plugin[0] : plugin))
+          .find(plugin => plugin === 'flow' || plugin === 'typescript');
```
Comment (Noviny) [inline packages/babel-plugin-extract-react-types/index.js]: Just trying to understand this correctly. Where is the plugin list derived from? Is it file specific?

Example problem is atlaskit contains both flow and typescript files. If both plugins exist, do we just get the first one?

**Comment by Noviny** [comment]
Oh, and a changeset would be ❤️

File: packages/babel-plugin-extract-react-types/index.js (hunk 1 above)
Comment (mitchellhamilton) [reply packages/babel-plugin-extract-react-types/index.js]: It's comes from the syntax plugins which are included in the flow and typescript presets. Both plugins can't exist together. I just checked atlaskit's webpack and babel configs and babel is only used on untyped/flow code whereas the typescript compiler is used on ts files so this babel plugin wouldn't work for ts unless you switched from the ts compiler to babel (at least on the website). To do that, you would probably want to use babel's overrides option and use the flow plugin on .js files and the typescript plugin on .ts and .tsx files.

**Comment by mitchellhamilton** [comment]
Just merged from master, fixed stuff and added a changeset

**Pull request merged**